### PR TITLE
Fixes zaproxy/zaproxy#1594 overhauled matching mechanism

### PR DESCRIPTION
**Pull request by kingthorin**

### CODE REVIEW

File: src/org/zaproxy/zap/extension/pscanrules/TestInfoSessionIdURL.java [hunk 1]
```diff
@@ -147,88 +152,56 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
...
-        String sessionIdValue;
-        String sessionIdName;
-        Pattern pattern;
-        Matcher matcher;
 
-        // Alternative implementation
-        // we search for all params looking for one session element
-        // maybe more efficient because we don't build Patterns any time
-        //String[] params = msg.getParamNames();
+        String uri = msg.getRequestHeader().getURI().toString();
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/pscanrules/TestInfoSessionIdURL.java]: The conversion to `String` could be delayed, to after the `urlParams` size check?

File: src/org/zaproxy/zap/extension/pscanrules/TestInfoSessionIdURL.java [hunk 2]
```diff
@@ -147,88 +152,56 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
...
 
-        // The Session ID list option param
+        TreeSet<HtmlParameter> urlParams = msg.getUrlParams();
+        
+        if (urlParams.size() == 0) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/pscanrules/TestInfoSessionIdURL.java]: Use `isEmpty()` instead?

File: src/org/zaproxy/zap/extension/pscanrules/TestInfoSessionIdURL.java [hunk 3]
```diff
@@ -147,88 +152,56 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
...
-                //kb = getKb().getString("sessionId/name");
-                //getKb().add("sessionId/name", sessionIdName);
-
-                if (sessionIdValue.length() > SESSION_TOKEN_MIN_LENGTH) { 
-	                
+        HttpSessionsParam sessionOptions = options.getParamSet(HttpSessionsParam.class);
+        List<String> sessionIds = sessionOptions.getDefaultTokensEnabled();
+        for (HtmlParameter param: urlParams) { //Iterate through the parameters
+        	//If the parameter name is one of those on the Session Token list from the options panel
+        	if (sessionIds.contains(param.getName())) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/pscanrules/TestInfoSessionIdURL.java]: The list `sessionIds` (by default) contains strings in lower case, parameter's name might not be.

**Comment by thc202** [comment]
Could you explain, in the commit message, the session's name comparison change? One has to read the diff to know what was the actual change.

**Comment by kingthorin** [comment]
Thanks, will work on addressing these items.

File: src/org/zaproxy/zap/extension/pscanrules/TestInfoSessionIdURL.java [hunk 4]
```diff
@@ -147,88 +152,56 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
...
-                //kb = getKb().getString("sessionId/name");
-                //getKb().add("sessionId/name", sessionIdName);
-
-                if (sessionIdValue.length() > SESSION_TOKEN_MIN_LENGTH) { 
-	                
+        HttpSessionsParam sessionOptions = options.getParamSet(HttpSessionsParam.class);
+        List<String> sessionIds = sessionOptions.getDefaultTokensEnabled();
+        for (HtmlParameter param: urlParams) { //Iterate through the parameters
+        	//If the parameter name is one of those on the Session Token list from the options panel
+        	if (sessionIds.contains(param.getName().toLowerCase())) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/pscanrules/TestInfoSessionIdURL.java]: Although it shouldn't make much difference in this case, consider converting to lower case in a locale independent way (i.e. `toLowerCase(Locale.ROOT)`).

**Comment by thc202** [comment]
There's a typo in the commit message (i.e. `machanism`).

**Comment by kingthorin** [comment]
Will do.

**Comment by kingthorin** [comment]
Done.

**Comment by thc202** [comment]
Thanks!

**Pull request merged**